solidity/contracts/contracts/QuantumGravityBridge.sol

```diff
@@ -0,0 +1,314 @@
+    bytes32 r;
+    bytes32 s;
+}
+
+/// @title Quantum Gravity Bridge: Celestia -> Ethereum, Data Availability relay.
+contract QuantumGravityBridge is OwnableUpgradeableWithExpiry {
+    // Don't change the order of state for working upgrades.
+    // AND BE AWARE OF INHERITANCE VARIABLES!
+    // Inherited contracts contain storage slots and must be accounted for in any upgrades
+    // always test an exact upgrade on testnet and localhost before mainnet upgrades.
+
+    ////////////////
+    // Immutables //
+    ////////////////
+
+    bytes32 public immutable BRIDGE_ID;
+
+    /////////////
+    // Storage //
+    /////////////
+
+    bytes32 public s_lastValidatorSetCheckpoint;
+    uint256 public s_powerThreshold;
+    uint256 public s_lastValidatorSetNonce;
+    uint256 public s_lastMessageRootNonce;
+    mapping(uint256 => bytes32) public s_messageRoots;
+
+    ////////////
+    // Events //
+    ////////////
+
+    /// @notice Emitted when a new root of messages is relayed.
+    /// @param nonce Nonce.
+    /// @param messagesRoot Merkle root of relayed messages.
+    event MessageRootEvent(uint256 indexed nonce, bytes32 messagesRoot);
+
+    /// @notice Emitted when the validator set is updated.
+    /// @param nonce Nonce.
+    /// @param powerThreshold New voting power threshold.
+    /// @param validatorSetHash Hash of new validator set.
+    event ValidatorSetUpdatedEvent(uint256 indexed nonce, uint256 powerThreshold, bytes32 validatorSetHash);
+
+    ////////////
+    // Errors //
+    ////////////
+
+    /// @notice Malformed current validator set.
+    error MalformedCurrentValidatorSet();
+
+    /// @notice Validator signature does not match.
+    error InvalidSignature();
+
+    /// @notice Submitted validator set signatures do not have enough power.
+    error InsufficientVotingPower();
+
+    /// @notice New validator set nonce must be greater than the current nonce.
+    error InvalidValidatorSetNonce();
+
+    /// @notice Supplied current validators and powers do not match checkpoint.
+    error SuppliedValidatorSetInvalid();
+
+    /// @notice Message root nonce nonce must be greater than the current nonce.
+    error InvalidMessageRootNonce();
+
+    ///////////////
+    // Functions //
+    ///////////////
+
+    /// @param _bridge_id Identifier of the bridge, used in signatures.
+    /// @param _powerThreshold Initial voting power that is needed to approve operations.
+    /// @param _validatorSetHash Initial validator set hash.
+    constructor(
+        bytes32 _bridge_id,
+        uint256 _powerThreshold,
+        bytes32 _validatorSetHash
+    ) {
+        __Ownable_init_unchained();
+
+        BRIDGE_ID = _bridge_id;
+
+        // CHECKS
+
+        uint256 nonce = 0;
+        bytes32 newCheckpoint = domainSeparateValidatorSetHash(_bridge_id, nonce, _powerThreshold, _validatorSetHash);
+
+        // EFFECTS
+
+        s_lastValidatorSetCheckpoint = newCheckpoint;
+        s_powerThreshold = _powerThreshold;
```

Comment:
Q: just for my understanding, `s_*` stands for self?

Reply:
It's a convention for denoting it's a state (i.e. storage) variable, since Solidity doesn't enforce at the compiler level some explicit distinction between storage and local variables. The original Peggo used `state_*`.

Reply:
> The original Peggo used state_*.

I like that one better as it's more obvious but OK.

Reply:
OK, here is another reason to use the original state_ prefix, the generated go code will read weirdly now: `SLastValidatorSetNonce` is very unclear vs. `StateLastValsetNonce` at least communicates that it this is about state. I'll change that back in the context of #7